types/evidence.go

```diff
@@ -245,6 +266,663 @@ func (dve *DuplicateVoteEvidence) ValidateBasic() error {
+		if !ok {
+			continue
+		}
+
+		if !valSet.HasAddress(sig.ValidatorAddress) {
+			evList = append(evList, &PhantomValidatorEvidence{
+				Header:                      alternativeHeader.Header,
+				Vote:                        alternativeHeader.Commit.GetVote(i),
+				LastHeightValidatorWasInSet: lastHeightValidatorWasInSet,
+			})
+		}
+	}
+
+	// If ValidatorsHash, NextValidatorsHash, ConsensusHash, AppHash, and
+	// LastResultsHash in alternativeHeader are different (incorrect application
+	// state transition), then it is a lunatic misbehavior => immediately
+	// slashable (#F5).
+	var invalidField string
+	switch {
+	case !bytes.Equal(committedHeader.ValidatorsHash, alternativeHeader.ValidatorsHash):
+		invalidField = "ValidatorsHash"
+	case !bytes.Equal(committedHeader.NextValidatorsHash, alternativeHeader.NextValidatorsHash):
+		invalidField = "NextValidatorsHash"
+	case !bytes.Equal(committedHeader.ConsensusHash, alternativeHeader.ConsensusHash):
+		invalidField = "ConsensusHash"
+	case !bytes.Equal(committedHeader.AppHash, alternativeHeader.AppHash):
+		invalidField = "AppHash"
+	case !bytes.Equal(committedHeader.LastResultsHash, alternativeHeader.LastResultsHash):
+		invalidField = "LastResultsHash"
+	}
+	if invalidField != "" {
+		for i, sig := range alternativeHeader.Commit.Signatures {
+			if sig.Absent() {
+				continue
+			}
+			evList = append(evList, &LunaticValidatorEvidence{
+				Header:             alternativeHeader.Header,
+				Vote:               alternativeHeader.Commit.GetVote(i),
+				InvalidHeaderField: invalidField,
+			})
+		}
+		return evList
+	}
+
+	// Use the fact that signatures are sorted by ValidatorAddress.
+	var (
+		i = 0
+		j = 0
+	)
+OUTER_LOOP:
+	for i < len(ev.H1.Commit.Signatures) {
+		sigA := ev.H1.Commit.Signatures[i]
+		if sigA.Absent() {
+			i++
+			continue
+		}
+		// FIXME: Replace with HasAddress once DuplicateVoteEvidence#PubKey is
+		// removed.
+		_, val := valSet.GetByAddress(sigA.ValidatorAddress)
+		if val == nil {
+			i++
+			continue
+		}
+
+		for j < len(ev.H2.Commit.Signatures) {
+			sigB := ev.H2.Commit.Signatures[j]
+			if sigB.Absent() {
+				j++
+				continue
+			}
+
+			switch bytes.Compare(sigA.ValidatorAddress, sigB.ValidatorAddress) {
+			case 0:
+				// if H1.Round == H2.Round, and some signers signed different precommit
+				// messages in both commits, then it is an equivocation misbehavior =>
+				// immediately slashable (#F1).
+				if ev.H1.Commit.Round == ev.H2.Commit.Round {
+					evList = append(evList, &DuplicateVoteEvidence{
+						PubKey: val.PubKey,
+						VoteA:  ev.H1.Commit.GetVote(i),
+						VoteB:  ev.H2.Commit.GetVote(j),
+					})
+				} else {
+					// if H1.Round != H2.Round we need to run full detection procedure => not
+					// immediately slashable.
+					evList = append(evList, &PotentialAmnesiaEvidence{
+						VoteA: ev.H1.Commit.GetVote(i),
+						VoteB: ev.H2.Commit.GetVote(j),
+					})
```

Comment:
Are we doing Potential Amnesia as a batch of votes? In this case we may want to change this then.

Reply:
Yes, but let's do it in a separate PR.

Reply:
Okay, I've appended it to the refactor list for the modules package